Documented the `#[repr(align(x))]` attribute.

bitshifter opened the pull request:
Updated reference for stabilization of the `#[repr(align(x))]` attribute. See tracking issue comments https://github.com/rust-lang/rust/issues/33626#issuecomment-348467804 for discussion.

Havvy commented:
I thought I removed most of the verbiage of the `repr` attribute from the attributes page. Whoops.

You'll instead want to add a section ```The `align` representation` to `type-layout.md`.

bitshifter commented:
Ah OK, should I leave the text I added to `attributes.md` since it's already talking about the other `repr` attributes?

Havvy commented:
It's better to just link to the full description of the representation attribute from the attributes.md page, so as to avoid duplicating information too much. I just forgot to clean up the attributes.md page when the type layout page landed.

Havvy commented on src/type-layout.md:
```diff
@@ -264,13 +264,23 @@ For all other enumerations, the layout is unspecified.
 Likewise, combining two primitive representations together is unspecified.
 
-### The `packed` Representation
+### `align(x)` Representation
```
This should be called ```The `align` Representation```.

Havvy commented on src/type-layout.md:
```diff
@@ -264,13 +264,23 @@ For all other enumerations, the layout is unspecified.
+two of type `u32`. The `align` representation cannot lower the alignment of a
+type.
+
+### `packed` Representation
```
You got rid of the `The` here.

Havvy commented on src/type-layout.md:
```diff
@@ -264,13 +264,23 @@ For all other enumerations, the layout is unspecified.
 It modifies the representation (either the default or `C`) by removing any
 padding bytes and forcing the alignment of the type to `1`.
 
+The `align` and `packed` representations cannot be applied on the same type and
```
This section should also be included in the align representation section.

Havvy commented on src/type-layout.md:
```diff
@@ -264,13 +264,23 @@ For all other enumerations, the layout is unspecified.
-### The `packed` Representation
+### `align(x)` Representation
+
+The `align(x)` representation can be used on `struct`s and `union`s to raise the
```
The paragraph structure here should mirror the packed representation section. So which categories of types it can be applied to should be one paragraph, and what it does should be another.

Havvy replied:
Remove the `(x)` from the name. Instead, describe how it takes the alignment as a parameter in the paragraph describing how it works.

bitshifter commented:
Thanks @Havvy I think I've addressed your feedback. This stuff hasn't been stabilized quite yet so you might want to wait until that happens before accepting the PR I guess.

Havvy commented:
Github shows that you did a force push, but I don't actually see any changes in the new commit. I even checked your branch and there's no difference there either.

bitshifter commented:
Whoops, I did a `git commit --amend` but forgot to add first. Should be updated now.

Havvy commented:
> I have not updated the reference Attribute syntax grammar documentation as the new syntax is only used by #[repr(align)] and I thought it might be pretty confusing to include a grammar which is only used by one kind of attribute for now. &mdash @bitshifter; [rust-lang/rust #33626](https://github.com/rust-lang/rust/issues/33626#issuecomment-353904791)

Even if it's a change used by one attribute, it should still be in the attribute grammar. Do you want to add it? If not, I'll do so in a separate PR.

bitshifter commented:
Sorry, I didn't see your comment about the grammar. I can look at adding it.

bitshifter commented:
Updated the grammar, hopefully I got it right! It looks like my stabilization PR is almost in.

Havvy commented:
@bitshifter There's a merge conflict. 😢

bitshifter commented:
@Havvy the conflict was on a section on repr attributes that I removed as part of this review, had recently been updated by @yellow-apple. Thus I've removed the edit with the rest of the section, but perhaps @yellow-apple should check the `type-layout.md` section on `#[repr(C)] Enums` to make sure it contains the same information (it's worded differently but seems semantically similar to me).

Havvy commented:
💟 Thanks!
